src/System.Text.Json/src/System/Text/Json/Node/JsonNumber.cs

```diff
@@ -0,0 +1,610 @@
+
+        /// <summary>
+        ///   Converts an Int32 to a JSON number.
+        /// </summary>
+        /// <param name="value">The value to convert.</param>
+        public static implicit operator JsonNumber(int value) => new JsonNumber(value);
+
+        /// <summary>
+        ///   Converts an Int64 to a JSON number.
+        /// </summary>
+        /// <param name="value">The value to convert.</param>
+        public static implicit operator JsonNumber(long value) => new JsonNumber(value);
+
+        /// <summary>
+        ///    Converts a Single to a JSON number.
+        /// </summary>
+        /// <param name="value">The value to convert.</param>
+        public static implicit operator JsonNumber(float value) => new JsonNumber(value);
+
+        /// <summary>
+        ///    Converts a Double to a JSON number.
+        /// </summary>
+        /// <param name="value">The value to convert.</param>
+        public static implicit operator JsonNumber(double value) => new JsonNumber(value);
+
+        /// <summary>
+        ///    Converts a SByte to a JSON number.
+        /// </summary>
+        /// <param name="value">The value to convert.</param>
+        [CLSCompliant(false)]
+        public static implicit operator JsonNumber(sbyte value) => new JsonNumber(value);
+
+        /// <summary>
+        ///    Converts a UInt16 to a JSON number.
+        /// </summary>
+        /// <param name="value">The value to convert.</param>
+        [CLSCompliant(false)]
+        public static implicit operator JsonNumber(ushort value) => new JsonNumber(value);
+
+        /// <summary>
+        ///    Converts a UInt32 to a JSON number.
+        /// </summary>
+        /// <param name="value">The value to convert.</param>
+        [CLSCompliant(false)]
+        public static implicit operator JsonNumber(uint value) => new JsonNumber(value);
+
+        /// <summary>
+        ///    Converts a UInt64 to a JSON number.
+        /// </summary>
+        /// <param name="value">The value to convert.</param>
+        [CLSCompliant(false)]
+        public static implicit operator JsonNumber(ulong value) => new JsonNumber(value);
+
+        /// <summary>
+        ///    Converts a Decimal to a JSON number.
+        /// </summary>
+        /// <param name="value">The value to convert.</param>
+        public static implicit operator JsonNumber(decimal value) => new JsonNumber(value);
+
+        /// <summary>
+        ///   Compares <paramref name="obj"/> to the numeric value of this instance. 
+        /// </summary>
+        /// <param name="obj">The object to compare against.</param>
+        /// <returns>
+        ///   <see langword="true"/> if the numeric value of this instance matches <paramref name="obj"/>,
+        ///   <see langword="false"/> otherwise.
+        /// </returns>
+        public override bool Equals(object obj) => obj is JsonNumber number && _value == number._value;
```

Comment:
Is this right? I guess that this would mean that if I have a JsonNumber(0) and a JsonNumber("0.0") this would return false? same for JsonNumber(0.0) and JsonNumber("0.00").

Reply:
I brought up a similar concern since the API comment makes it seem like the values are mathematically equivalent: https://github.com/dotnet/corefx/pull/39917#discussion_r309510072

Given our underlying representation is string in the current design, numeric values that are equivalent mathematically will not be equal here.

Looks like Newtonsoft.Json's `JValue` does quite a sophisticated equality under the covers:
https://github.com/JamesNK/Newtonsoft.Json/blob/7b8c3b0ed0380cf76d66894e81bf4d4d5b0bd796/Src/Newtonsoft.Json/Linq/JValue.cs#L834
https://github.com/JamesNK/Newtonsoft.Json/blob/7b8c3b0ed0380cf76d66894e81bf4d4d5b0bd796/Src/Newtonsoft.Json/Linq/JValue.cs#L246

But it too treats these as not equal (since they mismatch on the type).

However, it does treat 1.1e1 and 1.10e1 as equal:
```C#
JObject o = JObject.Parse(@"{'v1': 0, 'v2': 0.0, 'v3': '0.0'}");
JToken tok1 = o["v1"];
JToken tok2 = o["v2"];
JToken tok3 = o["v3"];
Assert.Equal(tok1, tok2);
Assert.Equal(tok2, tok3);

JValue v1 = (JValue)o["v1"];
JValue v2 = (JValue)o["v2"];
JValue v3 = (JValue)o["v3"];
Assert.False(v1.Equals(v2)); // v1 = Type Integer
Assert.False(v2.Equals(v3)); // v2 = Type Float
Assert.False(v3.Equals(v1)); // v3  = Type String

o = JObject.Parse(@"{'v1': 11, 'v2': 1.1e1, 'v3': 1.10e1}");
v1 = (JValue)o["v1"];
v2 = (JValue)o["v2"];
v3 = (JValue)o["v3"];
Assert.False(v1.Equals(v2)); // v1 = Type Integer
Assert.True(v2.Equals(v3)); // v2 == v3 = Type Float
```